File: android/app/src/main/java/com/designprototype/workshop/data/IdentityCardText.kt
package com.designprototype.workshop.data

/**
 * Turning what a text recogniser saw on an identity card into numbers worth offering a designer.
 *
 * ── WHY THIS FILE HAS NO ANDROID IN IT ────────────────────────────────────────────────────────
 *
 * ML Kit's recogniser cannot run in a JVM unit test — it is a native pipeline behind a Play Services
 * `Task`, and `android.jar` on the test classpath is stubs. So the recogniser is split in two: the
 * part that TALKS TO THE MODEL ([IdentityCardRecognizer]) and the part that DECIDES WHAT IS A NUMBER,
 * which is this file. This half is pure Kotlin over plain data and `IdentityCardTextTest` holds it to
 * every rule below without a device.
 *
 * Stated plainly, because a green test suite must not be read as more than it is: NOTHING HERE
 * EXERCISES THE RECOGNISER. Whether ML Kit reads a laminated card in courtyard light, and whether it
 * groups the twelve digits into one line, are hardware claims and this repository cannot make them.
 * What is verified here is what happens to the text once it exists.
 *
 * ── ONLY AADHAAR IS EXTRACTED ON THE DEVICE, AND THAT IS A FINDING RATHER THAN AN OMISSION ────
 *
 * An Aadhaar number can be recognised from raw text because it carries its own proof: exactly twelve
 * ASCII digits, never starting 0 or 1, and a Verhoeff check digit that catches every single-digit
 * misread and every adjacent transposition. Nothing else on the card satisfies that by accident.
 *
 * A Pehchan (PM Vishwakarma artisan ID) has NO checksum and no published fixed length — this repo's
 * own rule is "4 to 32 letters and digits", which also describes the artisan's name, the state code,
 * the scheme name and half the card. There is no way to pick it out of an unlabelled bag of strings.
 * The server does not do it either: `identity_ocr.pehchan_candidates` receives values a VISION MODEL
 * has already LABELLED as the Pehchan number and only normalises them. So an on-device Pehchan rule
 * would have to be invented here, against a card no one on this side has seen, with no checksum to
 * catch it being wrong — and a wrong Pehchan number is confirmed by a designer and stored, because
 * unlike a misread Aadhaar digit there is nothing downstream that can refuse it.
 *
 * The honest answer is therefore: the phone reads Aadhaar, the server reads Pehchan, and the panel
 * says which one answered. See `docs/DECISION-identity-card-ocr-on-android.md`. This becomes writable
 * the day somebody photographs a real Pehchan card and puts the text in a test here.
 *
 * ── NO LOGGING, EVER ──────────────────────────────────────────────────────────────────────────
 *
 * Every string this file handles is either an identity number or the text printed next to one. There
 * is no log statement at any level, and the rejected candidates are returned as a COUNT rather than
 * as values, for the same reason `DwIdentityOcrDto.rejectedAadhaarCount` is a count: a misread number
 * is still somebody's identity number.
 */
object IdentityCardText {

    /**
     * How much taller than the card's ordinary text a line must be to count as "printed large".
     *
     * The Aadhaar number is set in a noticeably larger face than the address and the parentage below
     * it, so height is real evidence about which twelve digits are THE number. The threshold is a
     * judgement, not a measurement — there is no card here to measure — and it is allowed to be one
     * ONLY because of the rule in [read]: layout evidence orders candidates and never admits or
     * excludes one. A threshold that is wrong costs the designer a second glance down the list; it
     * cannot cost them a wrong number.
     */
    private const val LARGE_LINE_RATIO = 1.25

    /**
     * One printed row a recogniser found, and the box it found it in.
     *
     * The box is OPTIONAL — a caller with no geometry passes nothing and every line reads as height
     * zero, at which case the size evidence simply switches itself off and ordering falls back to the
     * "on a line of its own" signal. That is what makes this type usable from a test that cares about
     * digits rather than about pixels.
     */
    data class RecognisedLine(
        val text: String,
        val top: Int = 0,
        val bottom: Int = 0,
    ) {
        /** Zero when no box was supplied, or when a recogniser returned a degenerate one. */
        internal val height: Int get() = (bottom - top).coerceAtLeast(0)
    }

    /**
     * What the device is willing to say about a photograph.
     *
     * [aadhaar] is BEST FIRST — see [read] for what "best" means and for why it is only an ordering.
     * [rejectedAadhaarCount] carries the same meaning as the server's field of that name: how many
     * distinct twelve-digit runs were found and failed their checks. It is the difference between
     * "the card was found and misread" (move to better light) and "nothing was found" (fill the
     * frame), which are different next actions for the designer.
     */
    data class IdentityCardReading(
        val aadhaar: List<String> = emptyList(),
        val rejectedAadhaarCount: Int = 0,
    )

    /**
     * Every Aadhaar number worth offering from [lines], best first, plus the rejected count.
     *
     * ── THE FILTER, WHICH IS THE WHOLE POINT ──────────────────────────────────────────────────
     *
     * A text recogniser hands back every numeral on the card: the VID, the enrolment number, the date
     * of birth, the pin code, the issue timestamp, the helpline. A digit string is not an Aadhaar
     * number, and offering a wrong-but-plausible one is worse than offering none, because the designer
     * confirms it and it becomes a deduplication key that belongs to nobody. So a run must survive all
     * four of:
     *
     *  1. it is a MAXIMAL run of digits-and-single-separators — see [scanDigitRuns];
     *  2. that run holds exactly twelve digits;
     *  3. it does not start 0 or 1, which UIDAI never issues;
     *  4. it satisfies the Verhoeff checksum ([ArtisanIdentity.verhoeffOk]).
     *
     * Rules 2-4 are `artisan_identity.aadhaar_error`'s three rules in the server's own order, so the
     * handset cannot offer a number the API would then refuse — the failure that leaves a designer
     * with a form they can neither submit nor correct, hours from the artisan whose card it was.
     *
     * ── RULE 1 IS DELIBERATELY STRICTER THAN THE SERVER, AND HERE IS THE CARD IT SAVES ────────
     *
     * The server matches `(?<![0-9])((?:[0-9][ \-]?){11}[0-9])(?![0-9])`, which accepts a twelve-digit
     * PREFIX of a longer grouped run. Every Aadhaar card also prints a sixteen-digit VID, grouped
     * 4-4-4-4. Against "9174 9963 5947 4681" that pattern matches "9174 9963 5947" — the lookahead
     * sees the SPACE after the twelfth digit, not the four digits after that — and yields a number
     * that is printed nowhere on the card. It then has to pass Verhoeff, which one arbitrary
     * twelve-digit string in ten does. The device therefore refuses to split a run: a group of digits
     * is offered only when the WHOLE run is twelve digits long.
     *
     * The device cannot defend against this on the server's behalf, either — a fabricated number only
     * reaches the client if it already passed Verhoeff, so re-checking it here cannot catch it. That
     * is a defect in `backend/app/services/identity_ocr.py` and is reported as one.
     *
     * ── AND THE ORDERING, WHICH ONLY ORDERS ───────────────────────────────────────────────────
     *
     * When more than one candidate survives, the one printed the way the card prints its number goes
     * first: on a line of its own, in a face larger than the surrounding text. EVERY survivor is still
     * returned and the panel still shows them all — this is the order they are read in, not a choice
     * made on the designer's behalf. Ranking that could DROP a candidate would be a silent decision
     * about which of two identity numbers a person gets, and no evidence available here is strong
     * enough to make that.
     */
    fun read(lines: List<IdentityCardText.RecognisedLine>): IdentityCardReading {
        // The median of the lines that actually carry a box. Median rather than mean because a card
        // photographed with a doorway in the frame produces a handful of enormous background lines,
        // and a mean would let them declare the real card text "small".
        val heights = lines.map { it.height }.filter { it > 0 }.sorted()
        val medianHeight = if (heights.isEmpty()) 0 else heights[heights.size / 2]

        val accepted = mutableListOf<Ranked>()
        var rejected = 0
        val seen = mutableSetOf<String>()

        lines.forEach { line ->
            scanDigitRuns(line.text).forEach { run ->
                val digits = run.digits
                // Not twelve digits: never a candidate, so never a REJECTION either. A date of birth
                // and a pin code are not failed readings of an Aadhaar number, and counting them
                // would tell the designer the card was misread when it was read perfectly.
                if (digits.length != ArtisanIdentity.AADHAAR_LENGTH) return@forEach
                // The same run read twice — the number and a second copy of it — is one candidate and
                // one rejection at most. Mirrors the server's `seen` set.
                if (!seen.add(digits)) return@forEach
                if (digits[0] == '0' || digits[0] == '1' || !ArtisanIdentity.verhoeffOk(digits)) {
                    rejected += 1
                    return@forEach
                }
                accepted += Ranked(
                    value = digits,
                    // The run is the whole of the line once whitespace is discounted. On the card the
                    // number sits on a line by itself; a twelve-digit run buried in a sentence is
                    // almost certainly something else that happens to be twelve digits long.
                    alone = line.text.removeRange(run.start, run.end).none { !it.isWhitespace() },
                    large = medianHeight > 0 && line.height >= medianHeight * LARGE_LINE_RATIO,
                )
            }
        }

        // Stable: candidates the evidence cannot separate stay in the order they were read, which is
        // top-to-bottom down the card. `sortedByDescending` is documented stable in Kotlin/JVM.
        return IdentityCardReading(
            aadhaar = accepted.sortedByDescending { it.score }.map { it.value },
            rejectedAadhaarCount = rejected,
        )
    }

    /** A surviving candidate with the layout evidence that decides where it sits in the list. */
    private data class Ranked(val value: String, val alone: Boolean, val large: Boolean) {
        /**
         * "On its own line" outweighs "printed large" because it is the more specific claim: plenty
         * of things on a card are set large, but only the number is set alone.
         */
        val score: Int get() = (if (alone) 2 else 0) + (if (large) 1 else 0)
    }

    /** A maximal digit run and where it sat in the line, so the rest of the line can be inspected. */
    private data class DigitRun(val digits: String, val start: Int, val end: Int)

    /**
     * Every MAXIMAL run of ASCII digits joined by at most one space or hyphen, in [text].
     *
     * Maximal is the load-bearing word: a run ends only where the next character is neither a digit
     * nor a single separator followed by a digit, so "9174 9963 5947 4681" is ONE run of sixteen
     * digits and cannot yield a twelve-digit candidate. See [read] for the card that costs.
     *
     * ASCII `'0'..'9'` ONLY, never `Char.isDigit`, which is Unicode-aware and returns true for the
     * Devanagari "१" and the fullwidth "２". Those would be carried through as-is and stored verbatim,
     * at which point the unique index that deduplicates artisans sees "१२३४५६७८९०१२" and
     * "123456789012" as two different people — the identical note sits over `aadhaarError` and over
     * the server's `_AADHAAR_RUN`, and it is the reason the Devanagari recognition model is NOT
     * shipped: nothing it could read would survive this line.
     */
    private fun scanDigitRuns(text: String): List<DigitRun> {
        val runs = mutableListOf<DigitRun>()
        var i = 0
        while (i < text.length) {
            if (text[i] !in '0'..'9') {
                i += 1
                continue
            }
            val start = i
            val digits = StringBuilder()
            var j = i
            while (j < text.length) {
                val c = text[j]
                when {
                    c in '0'..'9' -> {
                        digits.append(c)
                        j += 1
                    }
                    // A single separator counts as part of the run only when a digit follows it. Two
                    // spaces end the run, which is what keeps "1234  5678" two numbers rather than one
                    // — the same tolerance the server's `[ \-]?` allows, no more.
                    (c == ' ' || c == '-') && j + 1 < text.length && text[j + 1] in '0'..'9' -> j += 1
                    else -> break
                }
            }
            runs += DigitRun(digits.toString(), start, j)
            i = j
        }
        return runs
    }
}

File: frontend/lib/identityCardText.ts
/**
 * Turning RECOGNISED TEXT off an identity card into numbers worth offering — and refusing the rest.
 *
 * PURE ON PURPOSE. No React, no DOM, no `File`, no fetch. Everything that decides whether a
 * twelve-digit string is put in front of a designer lives here, so it can be tested in Node
 * (`e2e/identity-card-web-unit.spec.ts`) rather than through a browser and a camera. The browser
 * half — the recogniser, the bitmap, the feature detection — is `lib/identityCardLocal.ts` and it
 * makes no decisions at all.
 *
 * ── WHY THIS EXISTS AT ALL ────────────────────────────────────────────────────────────────────
 *
 * Until now the only reader was `POST /design-workshops/ocr/identity`, and the SERVER did this
 * parsing (`backend/app/services/identity_ocr.py::aadhaar_candidates`); the browser only re-checked
 * the candidates it was handed. A browser that recognises the text itself has raw text and nothing
 * else, so the same rule has to exist here — and "the same rule" has to mean the same rule, not a
 * paraphrase of it. See `docs/DECISION-identity-card-ocr-on-web.md` for how the two are kept in
 * step, and for why the ports are pinned by tests against the server's own fixtures rather than by
 * good intentions.
 *
 * ── A DIGIT STRING IS NOT AN AADHAAR NUMBER, AND THIS IS WHERE THAT IS ENFORCED ────────────────
 *
 * An Aadhaar card carries, in print, next to each other: the twelve-digit number, a SIXTEEN-digit
 * VID, a fourteen-digit enrolment number, a date, and a six-digit pin code. Any recogniser returns
 * all of them. Two filters stand between that and `Artisan.aadhaarNumber`, which is the
 * repository's deduplication key:
 *
 * 1. **The run must be a WHOLE digit token of exactly twelve digits.** This is the rule that is
 *    stricter here than the server's regex, and it was added because of a measurement rather than a
 *    hunch. The server matches `(?<![0-9])((?:[0-9][ \-]?){11}[0-9])(?![0-9])`, whose lookarounds
 *    only look at the immediately adjacent CHARACTER — so in the printed VID "9123 4567 8901 2345"
 *    the first twelve digits are followed by a SPACE, the lookahead is satisfied, and
 *    "912345678901" is extracted as an Aadhaar candidate. Verhoeff then rejects it about nine times
 *    in ten. Sampled over 200,000 Verhoeff-valid sixteen-digit numbers, **10.02% of them have a
 *    twelve-digit prefix that ALSO satisfies Verhoeff** — so roughly one card in ten would offer
 *    the front of its own VID as a number, and the confirm step cannot catch it: the panel prints
 *    "9123 4567 8901" grouped 4-4-4 and the card prints "9123 4567 8901 2345", so the first three
 *    groups match, which is exactly what a designer comparing group by group is looking for.
 *    Reading the maximal separator-joined token and demanding it be exactly twelve digits long
 *    refuses the VID, the enrolment number and the pin code outright, before any checksum runs.
 * 2. **Verhoeff, via the caller's validator.** `aadhaarProblem` is passed in — never imported —
 *    for the same reason `identityChoices` takes it: there is exactly one Aadhaar checksum on this
 *    client (`AadhaarField.aadhaarValidationError`) and this module does not get an opinion of its
 *    own about what a valid number is.
 *
 * A candidate that fails either filter is REFUSED, never offered with a warning. The rule the
 * whole feature rests on is that a misread digit produces twelve plausible digits belonging to
 * nobody, which collides with no one, passes every check the system has, and silently becomes a
 * second record for an artisan who is already in the repository.
 *
 * ── WHY THERE IS NO PEHCHAN PARSER HERE ───────────────────────────────────────────────────────
 *
 * The server's `pehchan_candidates` reads a STRUCTURED model reply: it asks a vision model for the
 * artisan ID and normalises what comes back. There is no equivalent to port. A PM Vishwakarma
 * artisan ID has no checksum and no fixed shape beyond "4–32 letters and digits", so extracting one
 * from free recognised text would nominate the artisan's name, the word "GOVERNMENT" and the
 * card's own serial with equal confidence, and the designer would be asked to confirm one of them
 * into a field that has no way of ever detecting the mistake. The local reader therefore offers
 * Aadhaar numbers only, and the surfaces that use it say so. The server route still returns Pehchan
 * candidates and `identityChoices` still filters them.
 *
 * ── NOTHING HERE LOGS A NUMBER ────────────────────────────────────────────────────────────────
 *
 * No console call, at any level, on any path. `rejectedCount` is a COUNT and never the values — a
 * rejected reading is still somebody's misread identity number, and it is also the one piece of
 * information that separates "the card was found and misread, take a better photograph" from "no
 * number was found at all, fill the frame".
 */

/**
 * A maximal digit token: digits joined by AT MOST ONE space or hyphen at a time.
 *
 * The separator alphabet is the server's, character for character — `[ \-]?` — so a card printed
 * "2345 6789 0124" and one printed "2345-6789-0124" are one token either way, while a newline, a
 * double space, a slash or a letter ends the token. Greedy by construction, which is the whole
 * point: the token is the LONGEST such run, so a sixteen-digit VID is one sixteen-digit token and
 * never a twelve-digit one with four digits left over.
 *
 * ASCII digits only, deliberately, and for the same reason `artisan_identity._ASCII_DIGITS` is:
 * `\d` under the `u` flag and `str.isdigit()` both accept Devanagari "१२३", which would sail
 * through every length and checksum test and then be STORED verbatim, at which point the unique
 * index sees two different strings for one person.
 */
const DIGIT_TOKEN = /[0-9](?:[ -]?[0-9])*/g;

/** Everything that is not an ASCII digit, for reducing a token to the digits it carries. */
const NON_DIGITS = /[^0-9]/g;

/** The one length an Aadhaar number has. Mirrors `AADHAAR_LENGTH` on the server and in the field. */
export const AADHAAR_LENGTH = 12;

export type IdentityTextReading = {
  /** Verhoeff-valid twelve-digit numbers, best first, de-duplicated. Safe to offer for confirmation. */
  aadhaar: string[];
  /**
   * How many twelve-digit tokens were read and then REFUSED — a count, never the values.
   *
   * Only whole twelve-digit tokens can be counted here. A VID, an enrolment number or a pin code is
   * not "a reading that failed its checksum" and must not inflate this, or the surface would tell a
   * designer that the card was misread and send them to photograph a perfectly good card again.
   */
  rejectedCount: number;
};

/**
 * Every Aadhaar number worth offering in `text`, plus how many twelve-digit readings were refused.
 *
 * `aadhaarProblem` returns null for an acceptable number and a sentence for a bad one — the exact
 * signature of `AadhaarField.aadhaarValidationError`, which every caller passes. It carries the
 * leading-0/1 rule and the Verhoeff checksum, so this function does not restate either.
 */
export function identityCandidatesFromText(
  text: string | null | undefined,
  aadhaarProblem: (digits: string) => string | null
): IdentityTextReading {
  const aadhaar: string[] = [];
  const seen = new Set<string>();
  let rejectedCount = 0;

  for (const match of (text ?? "").matchAll(DIGIT_TOKEN)) {
    const digits = match[0].replace(NON_DIGITS, "");
    // A token of any other length is not a misread Aadhaar number, it is a different number that
    // happens to be printed on the same card. It is refused here and counted nowhere.
    if (digits.length !== AADHAAR_LENGTH) continue;
    // The same number read twice — once off the front of the card and once off the back, or twice
    // in one frame — is one candidate and one rejection, not two of either.
    if (seen.has(digits)) continue;
    seen.add(digits);
    if (aadhaarProblem(digits)) {
      rejectedCount += 1;
      continue;
    }
    aadhaar.push(digits);
  }

  return { aadhaar, rejectedCount };
}

/**
 * "123456789012" -> "XXXX XXXX 9012" — the only form of a number any surface but the confirm panel
 * and the edit box may print.
 *
 * The same rule as `mask_aadhaar` on the server and `isMaskedIdentityNumber` on this client:
 * anything that is not a full number is masked ENTIRELY rather than partially revealed, so a
 * malformed value can never leak more than a well-formed one. Exported from here rather than from a
 * component so that error strings, counts and provenance lines have a masking function available
 * without importing React.
 */
export function maskIdentityNumber(value: string | null | undefined): string {
  const digits = (value ?? "").replace(NON_DIGITS, "");
  if (digits.length < 4) return "XXXX XXXX XXXX";
  return `XXXX XXXX ${digits.slice(-4)}`;
}

/**
 * "123456789012" -> "1234 5678 9012": the grouping the card prints and a person proofreads by.
 *
 * A dropped or doubled digit is visible in "1234 5678 901" in a way it is not in "123456789 01",
 * and a designer compares the card group by group. Only ever used where the full number is
 * legitimately on screen — the confirm button and the candidate readout.
 */
export function groupAadhaarForReading(digits: string): string {
  return digits.replace(/(\d{4})(?=\d)/g, "$1 ");
}
